src/python/pants/base/exception_sink.py

```diff
@@ -19,28 +23,78 @@
 class ExceptionSink(object):
   """A mutable singleton object representing where exceptions should be logged to."""
 
-  _destination = os.getcwd()
+  # TODO: see the bottom of this file where we call set_destination() and friends in order to
+  # properly setup global state.
+  # TODO: ???
+  _destination = None
+  # We need an exiter in order to know what to do after we log a fatal exception.
+  _exiter = None
+  # Where to log stacktraces to in a SIGUSR2 handler.
+  _trace_stream = None
+
+  _all_destinations = []
 
   def __new__(cls, *args, **kwargs):
     raise TypeError('Instances of {} are not allowed to be constructed!'
                     .format(cls.__name__))
 
+  class ExceptionSinkError(Exception): pass
+
+  # TODO (in response to the above): we can avoid a single entry point and make everything work with
+  # great ease if we can ensure all set_* methods are idempotent!
   @classmethod
   def set_destination(cls, dir_path):
-    cls._destination = dir_path
+    cls._destination = cls._check_or_create_new_destination(dir_path)
+    cls._all_destinations.append(cls._destination)
 
   @classmethod
   def get_destination(cls):
     return cls._destination
 
   @classmethod
-  def exceptions_log_path(cls, for_pid=None):
+  def exceptions_log_path(cls, for_pid=None, in_dir=None):
+    if not in_dir:
+      in_dir = cls.get_destination()
     intermediate_filename_component = '.{}'.format(for_pid) if for_pid else ''
     return os.path.join(
-      cls.get_destination(),
+      in_dir,
       'logs',
       'exceptions{}.log'.format(intermediate_filename_component))
 
+  @classmethod
+  def set_exiter(cls, exiter):
+    assert(isinstance(exiter, Exiter))
+    cls._exiter = exiter
+    sys.excepthook = cls._log_unhandled_exception_and_exit
+
+  @classmethod
+  def set_trace_stream(cls, trace_stream):
+    # TODO: validate trace_stream somehow!
+    # TODO: do we need to keep a reference to this? I think not.
+    cls._trace_stream = trace_stream
+
+    # TODO: move /all/ signal handling into a SignalHandler class defined in this file which is not
+    # a singleton and can be subclassed (like Exiter). Since signals are global, it is useful to
+    # be able to look at all the signal handlers which are set in one place.
+    if faulthandler.is_enabled():
+      faulthandler.disable()
+    faulthandler.enable(trace_stream)
+    # This permits a non-fatal `kill -31 <pants pid>` for stacktrace retrieval.
```

Comment:
The reason I didn't consider this such a huge change when we talked was I think because the biggest change I see is just "having an actual signal handler for SIGUSR2 instead of using `faulthandler`", and the code I threw away was already doing that in order to handle SIGINT and SIGQUIT. I'm not sure now whether the handling of those signals should be done in this PR. We currently set handlers for those signals in RemotePantsRunner to send a signal to the pantsd-runner process on control-c.

Reply:
Well, the biggest change is figuring out where to write in a signal handler as described in your comment, but after that's figured out.

Reply:
> having an actual signal handler for SIGUSR2 instead of using faulthandler

If you haven't already, take a look at the restrictions on signal handlers: https://docs.python.org/3/library/faulthandler.html ...

> The fault handler is called on catastrophic cases and therefore can only use signal-safe functions (e.g. it cannot allocate memory on the heap). Because of this limitation traceback dumping is minimal compared to normal Python tracebacks:

Reply:
That's extremely helpful! As an aside, I have used the [glibc docs on signal handlers](https://www.gnu.org/software/libc/manual/html_node/Signal-Handling.html) before and found the bit on re-entrancy and signal handlers to be super illuminating to understand what's allowed in a signal handler and applicable to other languages. I will peruse the python docs you linked because I was looking for that exactly and literally have a tab open to that page but just was only looking for method names and didn't read it. Thanks!